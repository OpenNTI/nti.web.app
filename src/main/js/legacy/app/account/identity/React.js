const Ext = require('@nti/extjs');
const { getService } = require('@nti/web-client');
const { Identity } = require('@nti/web-profiles');
const LoginActions = require('internal/legacy/login/Actions');
require('internal/legacy/overrides/ReactHarness');

const AccountActions = require('../Actions');

module.exports = exports = Ext.define(
	'NextThought.app.account.identity.React',
	{
		extend: 'NextThought.ReactHarness',
		alias: 'widget.identity-react',
		component: Identity.Session,

		setTheme(scope, knockout) {},

		initComponent: function () {
			this.callParent(arguments);
			Object.assign(this.initialConfig, {
				addHistory: true,
				baseroute: '/app',
			});
			this.AccountActions = AccountActions.create();
			this.LoginActions = LoginActions.create();
		},

		async afterRender() {
			this.callParent(arguments);
			const service = await getService();
			const supportLinks = service.getSupportLinks();

			this.mon(Ext.getBody(), {
				click: this.onBodyClicks.bind(this, supportLinks),
			});
		},

		onBodyClicks(supportLinks, e) {
			let action = e.getTarget('a[data-action]')?.dataset.action;
			const link = e.getTarget('a[href]');
			const href = link?.getAttribute('href');

			// This is used to launch support logic the same way everywhere
			// we have a link that goes to the supportContact url. (so we can
			// just emit that url in the dom and not worry about coping logic)
			if (
				supportLinks.supportContact === href &&
				supportLinks.internalSupport
			) {
				action = action || 'support';
			}

			if (action) {
				if (call(this, `do-${action}`, link, e) !== false) {
					e.stopEvent();
				}
			}
		},

		showLink(target) {
			target.setAttribute('target', '_blank');
			return false;
		},

		'do-welcomeGuide'(target) {
			this.AccountActions.showWelcomePage(target.href);
		},

		'do-about': 'showLink',

		'do-privacy'(target) {
			this.AccountActions.showPrivacy(target);
		},

		'do-privacyForMinors'(target) {
			this.AccountActions.showChildrensPrivacy(target.href);
		},

		'do-terms'(target) {
			this.AccountActions.showTermsOfService(target.href);
		},

		'do-support'(target) {
			this.AccountActions.showContactUs();
		},

		'do-helpSite': 'showLink',

		'do-impersonate'() {
			this.LoginActions.handleImpersonate();
		},

		'do-logout'() {
			this.LoginActions.handleLogout();
		},
	}
);

function call(scope, action, ...args) {
	if (typeof action === 'string') {
		return call(scope, scope[action], ...args);
	}

	return action.apply(scope, args);
}
