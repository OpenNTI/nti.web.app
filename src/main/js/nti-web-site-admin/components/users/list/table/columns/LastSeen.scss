@import '~@nti/style-common';

.users-table-container {
	.lastseen-col {
		color: $secondary-grey;
		font-size: 14px;
		line-height: 60px;
		width: 7rem;
	}
}
