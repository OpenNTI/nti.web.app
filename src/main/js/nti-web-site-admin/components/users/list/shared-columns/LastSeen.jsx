import React from 'react';

import { scoped } from '@nti/lib-locale';
import { DateTime, Typography, Variant } from '@nti/web-core';

const t = scoped('nti-web-site-admin.users.list.shared-columns.LastSeen', {
	title: 'Last Active',
	never: 'Never',
	now: 'Now',
});

const Type = {
	type: 'body',
};

LastSeenColumn.Name = t('title');
LastSeenColumn.CssClassName = css`
	width: 7rem;
`;

LastSeenColumn.SortOn = 'lastSeenTime';

LastSeenColumn.Create = props => Variant(LastSeenColumn, props);

export function LastSeenColumn({ item, getUser = x => x }) {
	const user = getUser(item);
	const lastSeenTime = user.getLastSeenTime?.();

	if (!lastSeenTime || lastSeenTime === 0) {
		return <Typography {...Type}>{t('never')}</Typography>;
	}

	const diff = Date.now() - lastSeenTime;

	if (diff <= 60 * 1000) {
		return <Typography {...Type}>{t('now')}</Typography>;
	}

	return <DateTime.RelativeAdverb date={lastSeenTime} {...Type} />;
}
