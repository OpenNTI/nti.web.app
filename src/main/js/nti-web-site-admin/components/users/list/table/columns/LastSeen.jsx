import React from 'react';
import PropTypes from 'prop-types';
import {scoped} from '@nti/lib-locale';
import {DateTime} from '@nti/web-commons';
import cx from 'classnames';

const t = scoped('nti-web-site-admin.components.users.list.table.columns.LastSeen', {
	never: 'Never',
	duration: '%(duration)s ago',
	title: 'Last Seen'
});


export default class LastSeen extends React.Component {
	static propTypes = {
		item: PropTypes.object.isRequired
	}

	static cssClassName = 'lastseen-col';

	static Name = () => t('title')

	static SortKey = 'lastSeenTime'

	render () {
		const {item} = this.props;

		const lastLoginTime = item.getLastLoginTime && item.getLastLoginTime();
		const lastSeenTime = item.getLastSeenTime && item.getLastSeenTime();
		let lastSeen = lastSeenTime || lastLoginTime;

		let label = t('never');

		if(lastSeen && lastLoginTime && lastLoginTime.getTime() > 0) {
			label = t('duration', { duration: DateTime.getNaturalDuration(Date.now() - lastSeen, 1)});
		}

		return <div className={cx('cell')}>{label}</div>;
	}
}
