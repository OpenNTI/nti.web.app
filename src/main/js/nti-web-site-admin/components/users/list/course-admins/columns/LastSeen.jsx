import { LastSeenColumn } from '../../shared-columns/LastSeen';

export const LastSeen = LastSeenColumn;
