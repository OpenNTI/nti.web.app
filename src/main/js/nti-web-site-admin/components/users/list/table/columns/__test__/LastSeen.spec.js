/* eslint-env jest */
import React from 'react';
import { create, act } from 'react-test-renderer';

import LastSeen from '../LastSeen';

describe('Site admin user table last seen column test', () => {
	const verifyColumn = date => {
		let cmp;
		act(() => {
			cmp = create(
				<LastSeen
					item={{
						getLastSeenTime: () => {
							return date;
						},
					}}
				/>
			);
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	};

	test('Test date is null', () => {
		// null value should show "Never"
		verifyColumn(null);
	});

	test('Test date is within past 60 seconds', () => {
		// within past 60 seconds should just show "Now"
		verifyColumn(new Date(Date.now() - 20000));
	});

	test('Test date is 100 seconds ago', () => {
		// More than 60 seconds ago, show in minute or hour granularity.. "1 minute ago" in this case
		verifyColumn(new Date(Date.now() - 100 * 1000));
	});
});
