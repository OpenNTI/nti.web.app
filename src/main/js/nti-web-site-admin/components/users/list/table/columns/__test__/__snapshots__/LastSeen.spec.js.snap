// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site admin user table last seen column test Test date is 100 seconds ago 1`] = `
<div>
  1 Minute ago
</div>
`;

exports[`Site admin user table last seen column test Test date is null 1`] = `
<div>
  Never
</div>
`;

exports[`Site admin user table last seen column test Test date is within past 60 seconds 1`] = `
<div>
  Now
</div>
`;
